### PR TITLE
More NNI Search Optimization Fixes

### PR DESCRIPTION
**Pull request by davidrich27**
## Description

- Adds a variety of changes to speed up NNI Search.
  - Adds option to NNIEvalEngine to only partially recompute likelihoods.
  - Adds a scored NNIs map that maintains a sort by score, and faster methods for obtaining the min/max scoring NNIs.
  - Add "new" NNIs and scored NNIs, for an already filtered list for current iteration adjacent NNIs to rescore/reevaluate.
- Added python test that checks that NNI search results matches those results from commits.
- Added header that keeps software version in source code.  Reports the git commit, branch, and git tags that the version of the software was built on.
- Various other refactoring.

- Debugged issue that caused internal DAG NNI Scores and Proposed NNI Scores to change unexpectedly
  - Added optional monitoring into python nni search for changes in scores, PVs, branch lengths for debugging.
  - Added options to nni search: whether to initialize branch lengths based on pre-NNI or start at default branch length, whether to fix or optimize branch lengths already contained in the DAG.

## Tests

N/A

## Checklist:

* [x] Code follows our detailed [contribution guidelines](CONTRIBUTING.md)
* [x] `clang-format` has been run
* [x] TODOs have been eliminated from the code
* [ ] Comments are up to date, document intent, and there are no commented-out code blocks
* [ ] Issue branch has been squashed and rebased on main branch
* [ ] GitHub CI build on PR branch completed successfully

### CODE REVIEW

**Review by matsen — COMMENTED** [review]
Thanks for this, Dave!

This is going to take a little bit for me to get through, and I think that I might need a little tour from you. But here are some questions.

* "Adds option to NNIEvalEngine to only partially recompute likelihoods": can you provide a little detail about that? Which likelihoods and under what circumstances? Is this an approximation? Can you point me to this in the code?
* There is a lot of code marked "Reevaluate"-- this is all about re-evaluating rejected NNIs?
* I'm trying to understand what a "new NNI" is, e.g. in new_scored_nnis_. How long does a new NNI stay new?
* Do we expect this new code to give identical results to the old code, just faster? If so, have you checked this?

It may be easiest to just talk these over-- happy to do so on Monday.

File: src/subsplit_dag.hpp [hunk 1]
```diff
@@ -495,6 +495,43 @@ class SubsplitDAG {
...
+    void Reinit(const SubsplitDAG &dag) {
+      prv_node_count = dag.NodeCount();
+      prv_edge_count = dag.EdgeCountWithLeafSubsplits();
+      node_reindexer = Reindexer::IdentityReindexer(prv_node_count);
+      edge_reindexer = Reindexer::IdentityReindexer(prv_edge_count);
+      added_node_ids = NodeIdVector();
+      added_edge_idxs = EdgeIdVector();
+    }
+
+    ModificationResult ComposeWith(const ModificationResult other) {
```
Comment (matsen) [inline src/subsplit_dag.hpp:508]: Not a huge deal, but is there a reason these code blocks are sitting in the hpp file?

**Review by matsen — COMMENTED** [review]
> "NNIsToReevaluate" marks NNIs we want to consider adding to the DAG. Depending on the option chosen, this will either be strictly new adjacent NNIs or all adjacent NNIs (including previously rejected NNIs). 

If I understand correctly, this "Reevaluate" includes some new things and potentially some old things, right?

When you say "marks" do you mean returns?

File: src/subsplit_dag.hpp (hunk 1 above)
Comment (davidrich27) [reply src/subsplit_dag.hpp:508]: No, I can put that in the cpp file.

**Comment by davidrich27** [comment]
Yes, sorry.  I meant returns.  NNIsToReevaluate is determined by the `reevaluate_rejected_nnis_` flag, to either be the new NNIs or adjacent NNIs (new and old NNIs).  Rejected NNIs are functionally equivalent to the old adjacent NNIs, because they were not accepted during the previous iteration.

**Comment by matsen** [comment]
OK, great. I'll wait for you to make these little modifications then I'll add a few comments if that's OK.

File: src/tp_evaluation_engine.hpp
Comment (chrisjenningsshaffer) [inline src/tp_evaluation_engine.hpp:1]: Totally unimportant typo at line 248: dest_edge_0id instead of dest_edge_id

File: src/nni_engine.cpp
Comment (chrisjenningsshaffer) [inline src/nni_engine.cpp:1]: To get the next box checked for the PR: remove commented out code at lines 731, 735, 737, and 751

File: src/nni_engine.hpp
Comment (chrisjenningsshaffer) [inline src/nni_engine.hpp:1]: To get the next box checked for the PR: remove commented out code at line 189

File: src/tp_engine.cpp
Comment (chrisjenningsshaffer) [inline src/tp_engine.cpp:1]: To get the next box checked for the PR: remove commented out code at lines 1100, 1101, 1111, 1198, and 1199

File: src/tp_evaluation_engine.cpp
Comment (chrisjenningsshaffer) [inline src/tp_evaluation_engine.cpp:1]: To get the next box checked for the PR: remove commented out code at lines 468, 470, 924, 1044, 1065, 1066

File: src/pybito.cpp [hunk 2]
```diff
@@ -990,9 +1035,17 @@ PYBIND11_MODULE(bito, m) {
...
-           py::arg("top_n"), py::arg("max_is_best") = true)
+           py::arg("top_k"), py::arg("max_is_best") = true)
+      // Custom filters
+      .def("set_filter_init_function", &NNIEngine::SetFilterInitFunction)
+      .def("set_filter_pre_score_function", &NNIEngine::SetFilterPreScoreFunction)
+      .def("set_filter_score_loop_function", &NNIEngine::SetFilterScoreLoopFunction)
+      .def("set_filter_post_score_function", &NNIEngine::SetFilterPostScoreFunction)
+      .def("set_filter_evaluate_function", &NNIEngine::SetFilterEvaluateFunction)
+      .def("set_filter_evaluate_loop_function",
+           &NNIEngine::SetFilterEvaluateLoopFunction)
```
Comment (chrisjenningsshaffer) [inline src/pybito.cpp:1048]: Do we need to expose the types StaticFilterUpdateFunction, StaticFilterScoreLoopFunction, etc? Or is pybind smart enough to already know them?

File: src/pybito.cpp [hunk 3]
```diff
@@ -934,34 +974,39 @@ PYBIND11_MODULE(bito, m) {
            "Get scores from NNIs from previous iterations.")
       .def("iter_count", &NNIEngine::GetIterationCount,
            "Get number of iterations of NNI search run.")
+      .def("nnis_to_rescore", &NNIEngine::GetNNIsToRescore,
```
Comment (chrisjenningsshaffer) [inline src/pybito.cpp:977]: Let's move past_scored_nnis and nnis_to_rescore up a few lines to the access section and add the relevant count functions here. Also let's add the count function for the scored nnis.

File: src/subsplit_dag.cpp
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/subsplit_dag.cpp:1]: Yes, removed.

File: src/nni_engine.cpp
Comment (davidrich27) [reply src/nni_engine.cpp:1]: Removed.

File: src/nni_engine.hpp
Comment (davidrich27) [reply src/nni_engine.hpp:1]: Removed.

File: src/bitset.cpp [hunk 4]
```diff
@@ -10,6 +10,8 @@
 
 #include "sugar.hpp"
 
+Bitset::Bitset() : value_(0, false) {}
```
Comment (matsen) [inline src/bitset.cpp:13]: Would you mind telling me where it was important to have a default constructor?

File: src/nni_engine.hpp [hunk 5]
```diff
@@ -263,58 +269,75 @@ class NNIEngine {
 
   // ** Filter Functions
 
-  // Initialization step for filter before beginning
+  // Initialization step for filter before first iteration.
   using StaticFilterInitFunction = std::function<void(NNIEngine &)>;
-  // Update step for any filter computation work to be done at start of each sweep.
+  // Function template for
```
Comment (matsen) [inline src/nni_engine.hpp:274]: truncated

File: src/nni_engine.hpp [hunk 6]
```diff
@@ -263,58 +269,75 @@ class NNIEngine {
...
+  // Function template for scoring to be performed on each adjacent NNI.
+  using StaticFilterScoreLoopFunction =
       std::function<double(NNIEngine &, const NNIOperation &)>;
   // Function template for processing an adjacent NNI to be accepted or rejected.
   using StaticFilterEvaluateFunction =
       std::function<void(NNIEngine &, const NNISet &, const NNIDoubleMap &,
                          const DoubleNNIPairSet &, NNISet &)>;
   using StaticFilterEvaluateLoopFunction =
       std::function<bool(NNIEngine &, const NNIOperation &, const double)>;
+  // Function template
```
Comment (matsen) [inline src/nni_engine.hpp:285]: complete please

File: src/pv_handler.hpp [hunk 7]
```diff
@@ -365,6 +367,34 @@ class PartialVectorHandler {
...
+        double val = pv_a(i, j);
+        if (val < min_val) {
+          min_val = val;
+        }
+      }
+    }
+    return min_val;
+  }
+
+  double Max(const PVId pvid_a) const {
```
Comment (matsen) [inline src/pv_handler.hpp:384]: Isn't there Eigen functionality for taking the min and the max?

File: test/nni_search.py [hunk 8]
```diff
@@ -0,0 +1,1454 @@
...
+class Results:
+    def __init__(self, args):
+        self.data_ = {}
+        self.pre_data_ = {}
+        self.df_ = None
+        self.args = args
+        pass
+
+    def data_init(self):
+        self.data_ = {
```
Comment (matsen) [inline test/nni_search.py:349]: Would this not be easier with a Pandas dataframe?

**Review by matsen — CHANGES_REQUESTED** [review]
As always, an impressive PR. A few things to fix here, and please look through Chris' review and make sure things are addressed/responded to. 

Thanks!

File: src/pybito.cpp (hunk 2 above)
Comment (davidrich27) [reply src/pybito.cpp:1048]: If you write python functions with the correct number of arguments, pybind can convert them to the C++ std::functions with the correct signature.

File: src/tp_engine.cpp
Comment (davidrich27) [reply src/tp_engine.cpp:1]: Removed.

File: src/tp_evaluation_engine.cpp
Comment (davidrich27) [reply src/tp_evaluation_engine.cpp:1]: Removed.

File: test/nni_search.py (hunk 8 above)
Comment (davidrich27) [reply test/nni_search.py]: Yeah that makes sense.  Converted to dataframe.

File: src/nni_engine.hpp (hunk 5 above)
Comment (davidrich27) [reply src/nni_engine.hpp]: Completed comment.

File: src/nni_engine.hpp (hunk 6 above)
Comment (davidrich27) [reply src/nni_engine.hpp]: Finished this comment as well.

File: src/pv_handler.hpp (hunk 7 above)
Comment (davidrich27) [reply src/pv_handler.hpp]: Yes, youre right.  I swapped with Eigen function.

File: src/pybito.cpp (hunk 3 above)
Comment (davidrich27) [reply src/pybito.cpp]: Moved functions and added count functions.

File: src/bitset.cpp (hunk 4 above)
Comment (davidrich27) [reply src/bitset.cpp]: Modified functions that required default constructor.

**Review by matsen — APPROVED** [review]
🎉

**Pull request merged**